### ISSUE TITLE
Packagization

### ISSUE THREAD

**Issue by odinuv** [keboola/sapi-python-client#3]
should be installable as a package

**Comment by pocin** [comment]
Before the package gets adopted I would suggest thinking about a proper name.
There are 2 that needs to be thought of (I think)

First, the package name

`from sapiclient import ...` is imo not very descriptive

I'd suggest at least prefixing it with kbc or keboola, so something like:

- kbcstorage (`from kbcstorage import Client`)
- kbcsapiclient (`from kbcsapiclient import Client`)
- kbcsapi (`from kbcsapi import Client`)
or something alike.

Second, the PyPi name (again `$ pip install sapiclient` doesn't feel right to me). 

Or are we ok with installing directly from github `$ pip install git+git://github.com/keboola/sapi-python-client.git` as in https://github.com/keboola/python-docker-application?
This can be same as the (importable) package name, but can be different.

How do you feel about 
- `pip install keboola-storage-client`
- `pip install kbc-storage-client`
- `pip install kbc-sapi-client`

Or any other suggestions?



On a related note - Does merging this sapi functionality with the https://github.com/keboola/python-docker-application makes sense?

`from keboola import StorageClient` would be great aswell as having all keboola related stuff under one package. (Are there any other components which would benefit from Python bindings?)

**Comment by pivnicek** [comment]
I'll respond from bottom to top:)

imho, this client is mainly intended for external python applications to interact with the keboola storage api, rather than keboola python docker applications, as our docker layer handles the storage interface for us in those cases.  (@odinuv do you agree with that assessment?)

for naming I really like `from keboola import StorageClient`

and I like `pip install keboola-storage-client` but am fine with installing directly via github too.

**Comment by pocin** [comment]
Just to be clear on one thing: for the `from keboola import StorageClient` to work we would have to either a) merge this with the python-docker-application, or b) rename the python-docker-application

I don' see any other way around this, since the python-docker-application already "owns" the keboola namespace


Ok, let's start with installing from GitHub directly and not complicate things with PyPi

**Comment by pivnicek** [comment]
Ah ic, yes of course, that's too bad.  That becomes a point in favour of the merge idea then.

**Comment by odinuv** [comment]
we definitely don't want to merge this into docker-application, they are two completely unrelated things. Sharing the namespace would be nice - but AFAIK, unless we want installing from pypi it's not a problem - i don't see a real difference between

**Comment by pocin** [comment]
I used the term merge in a wrong way. It definitely wouldn't make sense to directly merge it into the docker app, but maybe create a brand new "keboola-python-bindings" app (and keep the old docker existing to not break existing apps) which would contain both the docker and the StorageClient.

If not, then I am strongly against naming this package `keboola` (therefore no `from keboola import StorageClient`). I think this would bring only confusion and what would you do if you wanted to have installed both?

**Comment by odinuv** [comment]
so how about `kbc` ?

**Comment by pocin** [comment]
`kbc` sounds reasonable to me

Just to make sure we are on the same boat:

- The repo will be kept as is -> `https://github.com/keboola/sapi-python-client/`
- So far make it installable via `pip install git+git:///github.com/keboola/sapi-python-client.git`

And the directory structure will be something like this:
```
sapi-python-client
├── kbc
│   ├── __init__.py
│   ├── exceptions.py
│   ├── somemodule.py
│   └── storage.py
└── readme.MD
└── .git
└── .gitignore
```

which will allow

`from kbc import storage`

`from kbc.storage import StorageClient`

and if we do `from .storage import StorageClient` inside `sapi-python-client/kbc/__init__.py` we will be able to do just 
`from kbc import StorageClient`

Makes sense?  If so, I'll prepare a PR

**Comment by pivnicek** [comment]
seems reasonable enough to me 👍 ,  with tests and package docs that would make
```
sapi-python-client
├── doc
├── kbc
│   ├── __init__.py
│   ├── exceptions.py
│   ├── somemodule.py
│   └── storage.py
├── test
└── readme.MD
└── .git
└── .gitignore
```
and at that point we can throw it on travis

**Comment by odinuv** [comment]
wouldn't it be better to have just `client.py` ? So that we can have `metadata.py` and `components.py` in future like here https://github.com/keboola/storage-api-php-client/tree/master/src/Keboola/StorageApi

**Comment by odinuv** [comment]
come to think about it - it would be probably better to name the package `kbc_storage` - in the end it will be only storage API client (and they may be other clients, such as provisioning in future)

**Comment by pivnicek** [comment]
Good point, in that case we could rename the repo to `keboola/python-client` and then all keboola clients can live under a shared namespace of say `keboola_client`

**Comment by odinuv** [comment]
why? the clients are unrealated and pypi doesn't have real namespaces anyway

**Comment by pocin** [comment]
I am still quite new to kbc so I am not aware of all components, so take me with a grain of salt. @odinuv do I understand you correctly, that the kbc api has multiple components
- storage api for manipulating storage
- metadata api for ...
- components api for ..

and maybe sometimes in the future we'd like to develop py clients for all the components and you would prefer keeping the clients isolated is separate packages?

Honestly, I think it's more convinient (for both devs and users) to have them within one package

**Comment by pivnicek** [comment]
@odinuv I was thinking that the clients were related in so far as they consume keboola apis.  But essentially you're correct.

In that case I'd view it as `keboola_storage` for this one, and others could follow as `keboola_dev_portal` `keboola_provisioning` etc...  that's fine with me.

@pocin, the metadata and components @odinuv referred to are separate sections of the same storage api as you can see here http://docs.keboola.apiary.io/#reference/metadata and here http://docs.keboola.apiary.io/#reference/component-configurations respectively.
Having them as separate modules is just for ease of organisation.

**Comment by odinuv** [comment]
KBC has multiple APIs https://developers.keboola.com/overview/api/ - each API should have it separate package (if we ever make one), because they are completely unrelated.
Now we're dealing Storage API with http://docs.keboola.apiary.io/# which has multiple endpoints, which are quite unrelated. These endpoints are however part of the same client, because they share the same api (e.g. response formats, authentication) - so this must be part of one package

**Comment by odinuv** [comment]
so to me something like this would make most sense
`from kbc_storage_import Client`
`from kbc_storage_import Table`
`from kbc_storage_import Bucket`
`from kbc_storage_import Metadata`
`from kbc_storage_import Component`

or we can skip the `Client` altogether and refer directly to the API resources (which would be different to the PHP client, but probably more reasonable), or we can skip the `Table` and `Bucket` classes and be more close to the PHP client which has those historically included

**Comment by pivnicek** [comment]
> we can skip the Client altogether and refer directly to the API resources
I'm not sure how that would work, pass token directly to the resource class? 
```
table = Table("myToken")
tableData = table.loadData("in.c-myBucket.myTable")
```

Otherwise, I'd like to put a vote in for `keboola_storage` instead of `kbc_storage`.  kbc always makes me think of fried chicken:)

**Comment by pocin** [comment]
Oh boy that's a lot of api's. 

I think they are related by being part of the keboola connection. Imo that's why they should be within one package and each api should have it's module

so something like this
```
from kbc_whatever.storage import StorageClient
from kbc_whatever.runner import RunnerClient
from kbc_whatever.orchestrator import OrchestratorClient

```

I am not familiar with the api's at all so I might be wrong, but I guess they all would share more or less the same base class  from which all the other clients would inherit. 

So something along these lines
```
class BaseClient:
    def __init__(self, token):
        self.token = token
        self.root = 'https://connection.keboola.com/v2'
    def _get(endpoint):
        '''
        Make an authenticated GET request using self.token
        '''
        return requests.get(urljoin(self.root, endpoint), headers={'X-storage-token': self.token})


class StorageClient(BaseClient):
    def listBuckets():
        return self._get('storage/buckets?include=include')

    def list_buckets():
        return self.listBuckets()

class RunnerClient(BaseClient):
    pass
```

I can't imagine duplicating the `BaseClient` code in 12 different packages, one for each api (and installing 12 packages separately whenever I need some functionality)


AD this snippet

> so to me something like this would make most sense
>```
>from kbc_storage_import Client
>from kbc_storage_import Table
>from kbc_storage_import Bucket
>from kbc_storage_import Metadata
>from kbc_storage_import Component
>```

I dont think that makes much sense. Purely from the practical point of view, when instantiating the class (Client, Table, ...) you would have to supply the token each time, which is painfull.
```
table = Table('tokenasfasdfasdfsa')
table.list_tables() #????
```


----
**Those are my few cents, but you guys have more experience with the api so I will trust and respect your decision. I don't feel competent enough to have any final word.**

**Comment by pivnicek** [comment]
Yeah.  The thing is that the different APIs don't all use the same token types, headers, or even base uri, so other than being REST clients, they really share nothing further.

**Comment by odinuv** [comment]
when instantiating `Client`, you always have to supply the token, there is no way around, when instantiating other thing, we can supply `Client` instance

**Comment by odinuv** [comment]
it's should be split into different resources because if the storage api had a singe class, that class would have to have over 100 methods

**Comment by pocin** [comment]
Good point about the splitting.

It kinda makes sense to create an 
```
class Endpoint:
    def __init__(self, root, path, token):
        self.path = path

    def _get(**params):
        return requests.get(
            urllib.parse.urljoin(root, path),
            headers={'x-storage-token':self.token},
            params=params)


# either make a subclass
class BucketEndpoint(Endpoint):
    def listBuckets(self, *args, **kwargs):
        return self._get(*args, **kwargs)

class StorageClient:
    def __init__(self, token):
        self.token = token
        self.root = 'http://something.com'
        self.buckets = BucketsEndpoint(self.root, 'buckets'):
    
# and use like this:

client = StorageClient('apitoken')
client.buckets.listBuckets(id='1234')
```
Or something along the lines - This fixes both issues of avoiding 100 class method + passing of tokens, I think I've seen this design pattern in a https://github.com/charlesthk/python-mailchimp and it's convinient to use.

 but that's probably a little offtopic.

**Comment by odinuv** [comment]
that looks real good

**Comment by pivnicek** [comment]
great!

**Comment by Ogaday** [comment]
Hello! I recently began to develop my own Python Client for the Storage API (specifically related to Workspaces), not realising that this existed. I'd like to contribute, but don't want to redo existing work or clash with the direction of this project. Have you made any progress @pocin or @pivnicek ? The discussion you've had so far looks very good.

**Comment by pocin** [comment]
@Ogaday great to hear that! I have got some commits prepared, but I've been very busy lately. Honestly I am not completely convinced about making separate package for each api (I'd prefer one package encompassing all APIs). What are your thoughts on that?

Think about the boto3 client for managing aws, I think it's very similar problem. They have dozens of unrelated services/APIs  all  accessible from single package...

**Comment by Ogaday** [comment]
@pocin I'm not particularly opinionated about it. One the one hand I think I'd naturally lean towards one package per API to limit the scope of work, for separation of concerns and to do one thing and do it well with each package. On the other hand the boto3 client seems like a good example, and a single client for managing all of Keboola could be quite convenient. I can see how that would be attractive with regards to deduplicating work.

I'm happy to work on one package for all APIs.

**Comment by odinuv** [comment]
really let's keep it this to Storage API only, it itself is quite large. The other APIs do have completely different roles and even different authentication.

**Comment by pivnicek** [comment]
Welcome @Ogaday!

Regarding the packaging scope, it strikes me as a bridge that we can cross at a later date.  Let's build this out for the storage api for now.  If in the future there is another API that someone thinks should have a python client we can choose at that time to extend this package or make a different one.

**Comment by pocin** [comment]
As you wish guys, I'll send a PR (hopefully) this evening, I think I've got it almost ready. 

I just wanted to avoid confusing the users down the road. It's simple to design it well now rather than renaming and merging later on when the code is used in production. I can see this screw up in case of Microsoft services (which was keeping me busy and frustrated last week). They have tons of different applications, APIs, platforms and versions. They keep renaming and moving them around and it's just so difficult to learn it and use it.

If something like that happens to us, I will perform a harakiri (most likely disembowel myself with a keyboard)

**Comment by odinuv** [comment]
Will try to keep that in mind, we don't want to loose you :)

**Comment by Ogaday** [comment]
@pocin I'm keen to contribute to this but I don't want to duplicate work! Is there anything in particular that would be useful for me to look at until you make the pull request?

**Comment by pocin** [comment]
@Ogaday So sorry for the delay. I've made the PR along with a `dev` branch which now contains the refactored repo. I'd suggest checking it out, making PR's into it and once in a while we will merge it into the master. 

Let's decide what to do next in another issue #6?

**Comment by Ogaday** [comment]
@pocin no worries at all! Sounds perfect to me.

**Comment by pivnicek** [comment]
@pocin It's looking good!  I'm a little confused though, why not make the PR for master?

**Comment by pocin** [comment]
@pivnicek because the master should be a "stable" branch (it's what you see when visiting https://github.com/keboola/sapi-python-client/) + when you install the package via `pip install git+https://github.com/keboola/sapi-python-client`) (I believe) the master branch is installed by default. The PRs we will be making are not necessary complete in the sense of the whole package. So I dont think its a good idea to pollute the master branch with only half-working stuff. 

That's why I'd merge all PRs into the `dev` branch and merge the `dev` into `master` only once the `dev` branch is mature enough to be usable. 

But again, this is the first time I am collaborating with multiple folks on a project so I can easily be wrong :)

**Comment by pivnicek** [comment]
Ah, I see where you're coming from.

imho master should be the lead, and tagged release versions should be used as `stable`, `beta`, etc.  So, to get the current release it would be `git+https://github.com/keboola/sapi-python-client:0.01`

if you look at the releases tab here https://github.com/keboola/sapi-python-client/releases

Once we have `travis` and the tests in place this will become automatic.  For now though I think which ever way you are most comfortable with is fine.

**Comment by odinuv** [comment]
Let's not overengineer it. Any branch == unstable, `master` == stable. Whatever is stable goes to master - it does not have to be necessarily complete. The client won't support the complete API for long time (if ever), by far the most used features are creating and loading tables and creating buckets. Once we have those, we can have tag 1.0.0 . Then major features bump the version 1.1.0 etc. breaking changes bump the major version (though there shouldn't be much of them, the API is not changing). Whatever is in the master is the latest working (i.e. tests pass) and usable.

**Comment by Ogaday** [comment]
Sounds good to me. `master branch <=> stable` with versions controlled by git tags. #8 will dynamically track the git tag to provide version.

**Comment by odinuv** [comment]
y, I'll have a look at #8 in the afternoon

**Comment by odinuv** [comment]
done in c32af27f5e5e345471bc9f91000c4b2bacce71bc
